datalad_next/utils/shell_connection.py

```diff
@@ -0,0 +1,305 @@
+
+    Parameters
+    ----------
+    shell_cmd : list[str]
+        The command to execute the shell. It should be a list of strings that
+        is given to :func:`iter_subproc` as `args`-parameter. For example:
+        ``['ssh', '-p', '2222', 'localhost']``.
+
+    Yields
+    ------
+    :class:`ShellCommandExecutor`
+    """
+
+    def train(queue: Queue):
+        """Use a queue to allow chaining of iterables at different times"""
+        for iterable in iter(queue.get, None):
+            yield from iterable
+
+    subprocess_inputs: Queue = Queue()
+    with iter_subproc(shell_cmd,
+                      input=train(subprocess_inputs),
+                      bufsize=0) as shell_output:
+        cmd_executor = ShellCommandExecutor(subprocess_inputs, shell_output)
+        try:
+            # Skip initial login messages
+            result_0 = cmd_executor.execute_command(b'test')
+            for line in result_0:
+                lgr.debug('skipped login message line: %s', line)
+            if result_0.returncode != 1:
+                raise RuntimeError(
+                    'unexpected return code from command'
+                    f'`test`: {result_0.returncode}')
+
+            # Return the now ready connection
+            yield cmd_executor
+        finally:
+            # Ensure that ssh it terminated if an exception is raised by code
+            # that uses `ssh_connection`. This is necessary because
+            # `terminate`-call that is invoked by leaving the
+            # `iterable_subprocess`-context will not end the ssh-process. It
+            # will only terminate if stdin is closed, or if it is killed.
+            subprocess_inputs.put(None)
+
+
+class ShellCommandExecutor:
+    """Execute a command in a shell and read its result and return code
+
+    Objects of this class use interactive subprocess channels to:
+     - execute a single POSIX pipeline-list
+     - collect its output
+     - collect its return code
+
+    The command is extended by a postfix that prints a most likely unique
+    end marker and the return code of the command. The end marker is used to
+    detect the end of the output and the determine the return code of the
+    executed command.
+
+    Both, command output and command return code are available via
+    :class:`ShellCommandResponseGenerator`-instances that are returned by the
+    :meth:`execute_command`-method of this class.
+    """
+
+    def __init__(self,
+                 process_inputs: Queue,
+                 stdout: OutputFrom,
+                 ) -> None:
+        self.process_inputs = process_inputs
+        self.stdout = stdout
+        self.end_marker, self.command_postfix = self._get_marker()
+
+    def execute_command(self,
+                        command: bytes,
+                        *,
+                        stdin: Iterable[bytes] | None = None,
+                        ) -> ShellCommandResponseGenerator:
```

Comment:
I feel like this should be `__call__`. Would turn

```py
>>> with shell_connection(['bash']) as shell:
...     r = shell.execute_command(b'uname')
```

into

```py
>>> with shell_connection(['bash']) as shell:
...     r = shell(b'uname')
```

It would be nice (and convenient) to support `str`-type commands too. Not sure though about the implications (even now). Wouldn't they be interpreted given the target encoding used in the connected shell (on a possibly remote systems).

Reply:
Done in a869e0e96b6dec081e5d5018479239dc7fe09f39